sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala

```diff
@@ -114,4 +114,194 @@ object EstimationUtils {
     }
   }
 
+  /**
+   * Returns the number of the first bin into which a column values falls for a specified
+   * numeric equi-height histogram.
+   *
+   * @param value a literal value of a column
+   * @param bins an array of bins for a given numeric equi-height histogram
+   * @return the number of the first bin into which a column values falls.
+   */
+  def findFirstBinForValue(value: Double, bins: Array[HistogramBin]): Int = {
+    var binId = 0
+    bins.foreach { bin =>
+      if (value > bin.hi) binId += 1
```

Comment:
this looks more like a while loop pattern, can we use while loop here?

Reply:
Good point.  Actually while loop is better because it can exit early when the condition no longer qualifies.